"No FileSystem for scheme: file" inside Docker

Here's the full stack trace:

```
Exception in thread "main" java.io.IOException: No FileSystem for scheme: file
        at org.apache.hadoop.fs.FileSystem.getFileSystemClass(FileSystem.java:2660)
        at org.apache.hadoop.fs.FileSystem.createFileSystem(FileSystem.java:2667)
        at org.apache.hadoop.fs.FileSystem.access$200(FileSystem.java:94)
        at org.apache.hadoop.fs.FileSystem$Cache.getInternal(FileSystem.java:2703)
        at org.apache.hadoop.fs.FileSystem$Cache.get(FileSystem.java:2685)
        at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:373)
        at org.apache.spark.deploy.DependencyUtils$.resolveGlobPath(DependencyUtils.scala:191)
        at org.apache.spark.deploy.DependencyUtils$.$anonfun$resolveGlobPaths$2(DependencyUtils.scala:147)
        at org.apache.spark.deploy.DependencyUtils$.$anonfun$resolveGlobPaths$2$adapted(DependencyUtils.scala:145)
        at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:293)
        at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:36)
        at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:33)
        at scala.collection.mutable.WrappedArray.foreach(WrappedArray.scala:38)
        at scala.collection.TraversableLike.flatMap(TraversableLike.scala:293)
        at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:290)
        at scala.collection.AbstractTraversable.flatMap(Traversable.scala:108)
        at org.apache.spark.deploy.DependencyUtils$.resolveGlobPaths(DependencyUtils.scala:145)
        at org.apache.spark.deploy.SparkSubmit.$anonfun$prepareSubmitEnvironment$4(SparkSubmit.scala:363)
        at scala.Option.map(Option.scala:230)
        at org.apache.spark.deploy.SparkSubmit.prepareSubmitEnvironment(SparkSubmit.scala:363)
        at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:894)
        at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:180)
        at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:203)
        at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:90)
        at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:1039)
        at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:1048)
        at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
```
https://stackoverflow.com/questions/17265002/hadoop-no-filesystem-for-scheme-file has a lot of tips and I spent a lot of time taking apart JAR files. But this is all a red herring! The full logs have something more:

```
Caused by: java.lang.NullPointerException: null                                                                   
        at org.apache.hadoop.fs.Path.<init>(Path.java:104)                                                        
        at org.apache.hadoop.fs.Path.makeQualified(Path.java:440)                                                 
        at org.apache.hadoop.fs.FileSystem.makeQualified(FileSystem.java:469)                                     
        at org.apache.hadoop.fs.RawLocalFileSystem.getInitialWorkingDirectory(RawLocalFileSystem.java:557)        
        at org.apache.hadoop.fs.RawLocalFileSystem.<init>(RawLocalFileSystem.java:69)                             
        at org.apache.hadoop.fs.LocalFileSystem.<init>(LocalFileSystem.java:41)                                   
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)                                  
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)           
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)   
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)                                        
        at java.lang.Class.newInstance(Class.java:442)                                                            
        at java.util.ServiceLoader$LazyIterator.nextService(ServiceLoader.java:380)                            
        ... 30 common frames omitted
```

Here's an abridged `RawLocalFileSystem`:

```java
public class RawLocalFileSystem extends FileSystem {
  private Path workingDir;
  public RawLocalFileSystem() {
    workingDir = getInitialWorkingDirectory();
  }
  @Override
  public Path getWorkingDirectory() {
    return workingDir;
  }
  @Override
  protected Path getInitialWorkingDirectory() {
    return this.makeQualified(new Path(System.getProperty("user.dir")));
  }
  public Path makeQualified(Path path) {
    checkPath(path);
    return path.makeQualified(this.getUri(), this.getWorkingDirectory());
  }
}
```

That's a bit crazy! It uses `workingDirectory` in the constructor to initialize `workingDirectory`. Of course it gets a NullPointerException!

But if you read more of [`Path.makeQualified()`](https://github.com/apache/hadoop/blob/release-2.7.4-RC0/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/Path.java#L440) you see that line 440 that actually uses `workingDir` is only reached if the path we are looking at is not an absolute path. We are looking at `user.dir`. Why wouldn't that be absolute?

And yet line 440 is in the stack trace.

It turns out `user.dir` is `//lynxkite`. `Path` parses this as the path "" on host "lynxkite". And "" doesn't start with a slash, so.

### WHY IS `user.dir` SET TO `//lynxkite`???

It's set by our `stage/bin/lynxkite` script:

```bash
declare -r real_script_path="$(realpath "$0")"                                                                                                               
declare -r app_home="$(realpath "$(dirname "$real_script_path")")"
addJava "-Duser.dir=$(realpath "$(cd "${app_home}/.."; pwd -P)"  $(is_cygwin && echo "fix"))"                                                                
```

This script is generated automatically. Play, sbt-native-packager, and our `call_spark_submit.sh` script get combined in mysterious ways. This line comes from [Play](https://github.com/playframework/playframework/blob/2.8.7/dev-mode/sbt-plugin/src/main/scala/play/sbt/PlaySettings.scala#L206) actually.

But it doesn't seem wrong. `cd "${app_home}/.."; pwd -P` gives `/lynxkite`. And `realpath /lynxkite` in the shell gives `/lynxkite`. But inside the script, it gives `//lynxkite`!

Turns out the script has a function named `realpath` that behaves mostly the same as the real `realpath`, except if the path is a top level directory, it messes up and adds a slash. It comes from [`sbt-native-packager`](https://github.com/sbt/sbt-native-packager/blob/v1.9.4/src/main/resources/com/typesafe/sbt/packager/archetypes/scripts/bash-template#L12).

Guess I'll send them a pull request. For the short term I'm just replacing `user.dir=//` with `user.dir=/` after it's set.

--- a/tools/call_spark_submit.sh
+++ b/tools/call_spark_submit.sh
@@ -97,6 +97,8 @@ mode=${residual_args[0]}
 # -mem flag overrides KITE_MASTER_MEMORY_MB and we use 1024 if neither is set.
 final_app_mem=${app_mem:-${KITE_MASTER_MEMORY_MB:-1024}}
 final_java_opts="${java_opts} ${java_args[@]}"
+# Quick fix for https://github.com/lynxkite/lynxkite/issues/200.
+final_java_opts="${final_java_opts/user.dir=\/\//user.dir=\/}"
 
 # Cannot do this earlier, as the wrapping script is written in a -e hostile way. :(
 set -eo pipefail

